modules/core/src/main/scala/qq/droste/fixie.scala

```diff
@@ -0,0 +1,189 @@
+    import c.universe._
+    import Flag._
+    val inputs = annottees.map(_.tree).toList
+
+    require(inputs.length == 2, "@fixie should annotate a sealed [trait|abstract class] with companion")
+
+    val clait: ClassDef      = inputs.collect({ case c: ClassDef  => c }).head
+    val companion: ModuleDef = inputs.collect({ case c: ModuleDef => c }).head
+
+    val A: TypeName                     = c.freshName(TypeName("A"))
+    val claitTypeParams                 = clait.tparams
+    val claitTypeParamNames             = claitTypeParams.map(_.name)
+    val claitTypeParamNamesAsTrees      = claitTypeParams.map(x => Ident(x.name))
+    val claitTypeParamNamesWithA        = claitTypeParamNames :+ A
+    val claitTypeParamNamesWithAAsTrees = claitTypeParamNamesWithA.map(x => Ident(x))
+
+    val NonRecursiveAdtName: TypeName = TypeName(s"${clait.name}F")
+    val NonRecursiveAdtFullName: AppliedTypeTree =
+      AppliedTypeTree(Ident(NonRecursiveAdtName), claitTypeParamNamesWithAAsTrees)
+
+    def isSealed(classOrTrait: ClassDef): Boolean =
+      classOrTrait.mods.hasFlag(TRAIT) ||
+        classOrTrait.mods.hasFlag(ABSTRACT) &&
+          classOrTrait.mods.hasFlag(SEALED)
+
+    val isCase: PartialFunction[Tree, Tree] = {
+      case c: ClassDef if c.mods.hasFlag(CASE)  => c
+      case c: ModuleDef if c.mods.hasFlag(CASE) => c
+    }
+
+    val AdtCases = companion.impl.body.collect(isCase)
+
+    val convertCaseClassParam: ValDef => ValDef = { valDef =>
+      val recursive = valDef.tpt.toString == clait.name.toString || valDef.tpt.toString.contains(clait.name.toString)
+
+      if (recursive) {
+        ValDef(valDef.mods, valDef.name, Ident(A), valDef.rhs)
+      } else valDef
+    }
+
+    def convertCaseObject(module: ModuleDef): ClassDef = {
+      val Name: TypeName = TypeName(s"${module.name}F")
+
+      q"case class $Name[..${clait.tparams}, $A]() extends $NonRecursiveAdtFullName"
+    }
+
+    def getCaseClassParams(caseClass: ClassDef): List[ValDef] =
+      caseClass.impl.body
+        .collect({
+          case v: ValDef if v.mods.hasFlag(PARAMACCESSOR) && v.mods.hasFlag(CASEACCESSOR) => v
+        })
+        .map(convertCaseClassParam)
+
+    def convertCaseClass(caseClass: ClassDef): ClassDef = {
+      val Name: TypeName = TypeName(s"${caseClass.name}F")
+
+      val params: List[ValDef] = getCaseClassParams(caseClass)
+
+      q"case class $Name[..${clait.tparams}, $A](..$params) extends $NonRecursiveAdtFullName"
+    }
+
+    val NonRecursiveAdtCases: List[ClassDef] =
+      AdtCases.map {
+        case c: ModuleDef => convertCaseObject(c)
+        case c: ClassDef  => convertCaseClass(c)
+        case _            => sys.error("Nope")
+      }
+
+    val nonRecursiveAdt: ClassDef =
+      q"""sealed trait $NonRecursiveAdtName[..${clait.tparams}, $A] extends Product with Serializable"""
+
+    val λ: TypeDef = q"type λ[α] = $NonRecursiveAdtName[..${clait.tparams.map(_.name)}, α]"
+
+    val functorInstance = q"""
```

Comment:
Can we provide a `Traverse` instead?

Reply:
Currently kittens doesn't derive traverse. Only `Functor` and `Foldable` in that hierarchy